Single provider only

TehShrike opened the pull request:
Fixes #1 
Fixes #2 

As talked about in #1 and in chat, this is a PARADIGM SHIFT!

This change centers Mannish around decoupling function calls across parts of your application.

I tried to cover the new thought process in the updated readme.

ArtskydJ commented on readme.md:
```diff
@@ -1,57 +1,79 @@
+![](mannish.jpg)
 
-# Usage
+This is a fairly specific implementation of the mediator pattern - given a string name, it calls a function and returns a promise with the functions response.
```
"functions" should have an apostrophe, I think.

ArtskydJ commented on readme.md:
```diff
@@ -1,57 +1,79 @@
+			}
+		}
+	}
+}
```
Anyone trying to skim the readme would probably take this code block to be example usage, and then be confused.

TehShrike replied:
Yeah, you're right.  I should move it to a gist or just drop it entirely.

ArtskydJ commented on readme.md:
```diff
@@ -1,57 +1,79 @@
+## `mediator.provide(name, function)`
 
-## Publish a request for something
+Supplies a function to handle all calls to the given name.
```
It should be noted that you can return a promise, and provide the response in an async manner.

ArtskydJ commented on readme.md:
````diff
@@ -1,57 +1,79 @@
 ```
 
-The callback function is totally optional.  You can publish data objects to the app without any need for a response.
+## `promise = mediator.call(name, ...arguments)`
````
People who read code that uses `mannish` in the future, might confuse `mediator.call` to be invoking [`Function.prototype.call`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call)

TehShrike replied:
Hmm, I didn't think of that, that's a good point.

ArtskydJ commented on readme.md:
````diff
@@ -1,57 +1,79 @@
-```
+- Is it ever necessary to remove all providers?
+- Is it ever necessary to remove a single provider?
+	- If so, should anyone be able to remove the provider, or just the code that added it? (Should the mediator expose `removeProvider` or should `provide` return a removal function?)
````
Replacing a provider is something that people might want to do.

TehShrike replied:
Maybe so, but I couldn't think of a real-life use case so I wanted to hold off until it was necessary.

ArtskydJ commented on index.js:
```diff
@@ -1,70 +1,34 @@
-const denodeify = require('then-denodeify')
-const nodeify = require('then-nodeify')
+module.exports = function createMediator() {
+	const providers = Object.create(null)
```
Why not `const providers = {}`?

TehShrike replied:
So that it can be safely used as a map - normal objects have a prototype Object, which means that `mediator.call('constructor')` would totally do something even if you had't added a provider for it yet.

ArtskydJ commented on index.js:
```diff
@@ -1,70 +1,34 @@
+		} else if (typeof name !== 'string') {
+			throw new Error(`The provider name must be a string`)
+		} else if (providers[name]) {
+			throw new Error(`There is already a provider for ${name}`)
```
The error messages might be more readable with quote marks around `${name}`, if someone uses a multiple-word string as the name. (Lines 14 and 24.)

This is pretty opinionated, feel free to disregard. :)

TehShrike commented:
@ArtskydJ [raises a good point](https://github.com/TehShrike/mannish/pull/3#discussion_r100171826) that `call` could easily be confused with the method on the Function prototype.

Is that acceptable?  What name would be better?

ArtskydJ commented on test/test.js:
```diff
@@ -0,0 +1,95 @@
+test('Should work with multiple arguments', t => {
+	const mediator = mannish()
+
+	mediator.provide('taco', function(first, second) {
```
Copy & Paste is my guess.

TehShrike replied:
A good chunk of the tests were left over from when the module was first written, which is why the string quotes and function definitions aren't consistent everywhere.  I'll clean that up a bit.

ArtskydJ commented on test/test.js:
```diff
@@ -0,0 +1,95 @@
+	})
+})
+
+test(`Errors thrown in the provider should result in rejected responses`, t => {
```
A reason to not use ticks everywhere is the confusion that goes along with it. If I'm messing with someone's code base, I try to check what type of quotes they're using, and use those. I would be surprised by people using ticks if they don't have variables `${var}`, and it would looks like a mistake, even though it works fine.

Does it really reduce cognitive overhead? I think it's negligible. It does have the upside that you can use apostrophes and double quotes without escaping them. It's pretty rare that I want a tick in a string compared to apostrophes and double quotes.

Anyway, :+1: for consistency. :grin:

TehShrike commented on index.js:
```diff
@@ -1,70 +1,34 @@
-				unsubscribeFunctions.splice(index, 1)
+				return Promise.resolve(response)
+			} catch (err) {
+				return Promise.reject(err)
```
Mediator errors should be like SQL query syntax errors - they should cause loud obvious failures to be caught in development, and not dealt with by run-time code.

This particular error isn't technically a mediator error, it would be caused by run-time errors thrown by the provider function.

The other errors that the mediator is actually constructing itself, though - I don't think there would be any benefit to making them detectable.  I'm expecting them to bubble all the way back out to the developer.

In a perfect world that sort of error "you haven't yet set a provider for this function name" would be caught at compile-time, not run-time.

TehShrike commented on readme.md:
```diff
@@ -1,57 +1,60 @@
-appContext.publish('unixTimestamp', new Date(), function(err, result) {
-	// result is a ten-digit number, wheee
+mediator.provide('pie', recipe => {
+	Array.isArray(recipe.contents) // => true
```
That `Array.isArray` was actually more just to assert that the code was working the way that it should be when jsmd checks it for correctness.

I do really like the idea of making the example show both both possible return types, though.

TehShrike commented on test/test.js:
```diff
@@ -0,0 +1,95 @@
+const mannish = require('../')
+
+test('Basic functionality: should call the right function and get the response', t => {
+	t.plan(2)
```
Good catch.

TehShrike commented:
I think I've addressed the comments so far.  If none of the discussions get added to, I'll probably merge soon.

I think I can live with `mediator.call('functionName', arg)`, but I'm open to more discussion on that too.

ArtskydJ commented:
[Words that we could use instead of `call`.](http://www.thesaurus.com/browse/invoke) :grin:

------------

I wonder about support for swapping providers. Obviously if needed, it can be added later.

```js
mannish.provide('ajax', function (url) {
    var resultPromise = new Promise()
    xhr(url, function (err, res) {
        if (err) resultPromise.reject(err)
        else resultPromise.resolve(res)
    })
    return resultPromise
})

// somewhere else...

mannish.call('ajax', 'http://www.josephdykstra.com').then(res => console.log(res))

// later

if (typeof fetch !== 'undefined') {
    mannish.provide('ajax', fetch)
}

// now future mannish.call invocations will use `fetch`
```
This example actually proves nothing, because the two `mannish.provide` statements could have been merged. But maybe it's helpful nonetheless.

saibotsivad commented:
One other thing is: you shouldn't use images without attribution.

TehShrike commented:
Not even if I cite precedence via [glob](https://www.npmjs.com/package/glob)? :-D (they're both from Adventure Time)
